store/src/main/java/org/apache/rocketmq/store/CommitLog.java

```diff
@@ -815,89 +809,72 @@ public CompletableFuture<PutMessageResult> asyncPutMessage(final MessageExtBroke
             }
         }
 
-        topicQueueLock.lock(topicQueueKey);
-        try {
+        PutMessageThreadLocal putMessageThreadLocal = this.putMessageThreadLocal.get();
```

Comment:
Is there a race issue? I notice you use thread local variable out of the critical region.

Reply:
> 

Yes, we serialize outside of the lock, so each thread will have a ByteBuf for storing serialized data.